ibis/expr/operations.py

```diff
@@ -2438,6 +2438,37 @@ class TimestampDelta(ValueOp):
     output_type = rules.shape_like_arg(0, 'timestamp')
 
 
+# TODO: might be a single TemporalSubtract instead with TemporalBinaryPromoter
+class TimeSubtract(ValueOp):
+
+    input_type = [rules.time, rules.time]
+    output_type = rules.shape_like_arg(0, dt.Interval(unit='s'))
+
+
+class DateSubtract(ValueOp):
+
+    input_type = [rules.date, rules.date]
+    output_type = rules.shape_like_arg(0, dt.Interval(unit='d'))
+
+
+class TimestampSubtract(ValueOp):
+
+    input_type = [rules.timestamp, rules.timestamp]
+    output_type = rules.shape_like_arg(0, dt.Interval(unit='s'))
+
+
+class IntervalAdd(ValueOp):  # __radd__
+
+    input_type = [rules.temporal, rules.interval]
+    output_type = rules.shape_like_arg(0, 'timestamp')
```

Comment:
I think the output type here is a bit more complicated since `date + interval` doesn't always return a timestamp. You can either perform the checking in the `output_type` method or you can make three different add classes that directly express the output type. The latter is a bit easier to reason about IMO.